File: src/com/aslan/module/cipher/segments/SerialSegmentEncipher.js
import AbstractSegment from "./AbstractSetment"

export default class SerialSegmentEncipher extends AbstractSegment {

	run1(IN, offset, length) {
		let algorithm = this.algorithm;
		let key = this.key;
		let L = this.info.L;
		let N = this.info.N;
		let round = (length >> L) >> 3;
		let remainder = (length >> L) & 7;
		for (let i = 0; i < round; i++) {
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
		}
		for (let i = 0; i < remainder; i++) {
			algorithm.enc(new Uint8Array(IN, offset, N), key.update());
			offset += N;
		}
		return length;
	}

	run2(IN, inf, OUT, ouf, length) {
		let algorithm = this.algorithm;
		let padding = this.padding;
		let key = this.key;
		let N = this.info.N, fill = this.fill;
		let H = N - fill, F = N;
		let round = Math.floor(length / H) >> 2;
		let remainder = Math.floor(length / H) & 3;
		let buf;
		for (let i = 0; i < round; i++) {
			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, H));
			padding.padding(buf, H, fill);
			algorithm.enc(buf, key.update());
			inf += H;
			ouf += F;

			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, H));
			padding.padding(buf, H, fill);
			algorithm.enc(buf, key.update());
			inf += H;
			ouf += F;

			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, H));
			padding.padding(buf, H, fill);
			algorithm.enc(buf, key.update());
			inf += H;
			ouf += F;

			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, H));
			padding.padding(buf, H, fill);
			algorithm.enc(buf, key.update());
			inf += H;
			ouf += F;
		}
		for (let i = 0; i < remainder; i++) {
			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, H));
			padding.padding(buf, H, fill);
			algorithm.enc(buf, key.update());
			inf += H;
			ouf += F;
		}
		let lost = length % H;
		if (lost > 0) {
			buf = new Uint8Array(OUT, ouf, N);
			buf.set(new Uint8Array(IN, inf, lost));
			padding.padding(buf, lost, F - lost);
			algorithm.enc(buf, key.update());
		}
		return Math.ceil(1.0 * length / H) * F;
	}
}

File: src/com/aslan/module/cipher/segments/SerialSegmentEncipher.java
package com.aslan.module.cipher.segments;

public class SerialSegmentEncipher extends AbstractSegment {

    public int exec(byte[] in, int offset, int length) {
        int L = info.L, N = info.N;
        int round = (length >> L) >> 3;
        int remainder = (length >> L) & 7;
        for (int i = 0; i < round; i++) {
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
            algorithm.enc(in, offset, key.update());
            offset += N;
        }
        for (int i = 0; i < remainder; i++) {
            algorithm.enc(in, offset, key.update());
            offset += N;
        }
        return length;
    }

    public int exec(byte[] in, int inf, byte[] out, int ouf, int length) {
        int N = info.N;
        int H = N - fill, F = N;
        int round = (int) Math.floor(length / H) >> 2;
        int remainder = (int) Math.floor(length / H) & 3;
        for (int i = 0; i < round; i++) {
            System.arraycopy(in, inf, out, ouf, H);
            padding.padding(out, H, fill);
            algorithm.enc(out, ouf, key.update());
            inf += H;
            ouf += F;
            System.arraycopy(in, inf, out, ouf, H);
            padding.padding(out, H, fill);
            algorithm.enc(out, ouf, key.update());
            inf += H;
            ouf += F;
            System.arraycopy(in, inf, out, ouf, H);
            padding.padding(out, H, fill);
            algorithm.enc(out, ouf, key.update());
            inf += H;
            ouf += F;
            System.arraycopy(in, inf, out, ouf, H);
            padding.padding(out, H, fill);
            algorithm.enc(out, ouf, key.update());
            inf += H;
            ouf += F;
        }
        for (int i = 0; i < remainder; i++) {
            System.arraycopy(in, inf, out, ouf, H);
            padding.padding(out, H, fill);
            algorithm.enc(out, ouf, key.update());
            inf += H;
            ouf += F;
        }
        int lost = length % H;
        if (lost > 0) {
            System.arraycopy(in, inf, out, ouf, lost);
            padding.padding(out, lost, F - lost);
            algorithm.enc(out, ouf, key.update());
        }
        return (int) (Math.ceil(1.0 * length / H) * F);
    }
}